burotel/indico_burotel/tasks.py

```diff
@@ -0,0 +1,188 @@
+
+    # get all days between two dates
+    subq = db.session.query(db.func.generate_series(
+        reservation.start_dt,
+        now_utc(),
+        db.func.cast('1 day', Interval)
+    ).label('d')).subquery()
+
+    # count all days except for weekends (working days)
+    query = db.session.query(
+        db.func.count(subq.c.d).label('d')
+    ).filter(db.func.extract('dow', subq.c.d).notin_({0, 6})).subquery()
+
+    return reservation.query.filter(
+        reservation.state == ReservationState.pending,
+        # booking ends in the future
+        reservation.end_dt > now_utc(),
+        # booking has started more than `delta` working days ago
+        query.c.d > delta,
+        Room.attributes.any(
+            db.and_(
+                RoomAttributeAssociation.attribute == attr,
+                RoomAttributeAssociation.value.isnot(None)
+            )
+        )
+    ).join(Room)
+
+
+def _adams_request(plugin, action, user, room, start_dt, end_dt):
+    username = plugin.settings.get('adams_username')
+    password = plugin.settings.get('adams_password')
+    logger = plugin.logger
+
+    person_id = _find_person_id(plugin, user)
+
+    if not person_id:
+        logger.error("Task failed: can't find a Person ID for %s", user)
+        return
+
+    url = plugin.settings.get('adams_service_url').format(
+        action=action,
+        person_id=person_id,
+        room=room.name.replace('-', '/'),
+        start_dt=start_dt.isoformat(),
+        end_dt=end_dt.isoformat()
+    )
+
+    try:
+        res = requests.request('DELETE' if action == 'cancel' else 'POST', url, auth=(username, password), timeout=10)
+        res.raise_for_status()
+    except Timeout:
+        logger.warning('Request timed out')
+        raise Retry(message='Request timeout')
+    except RequestException as e:
+        logger.exception('Request failed (%d): %s', e.response.status_code, e.response.content)
+    else:
+        if res.status_code == 200:
+            return
+        logger.error('Unexpected response status code:\nURL: %s\nCode: %s\nResponse: %s',
+                     url, res.status_code, res.text)
+
+
+@celery.periodic_task(run_every=crontab(minute='0', hour='8'), plugin='burotel')
+def auto_cancel_bookings():
+    # bookings which should be cancelled
+    for booking in _build_query(3):
+        current_plugin.logger.info('Auto-cancelling booking {}'.format(booking))
+        booking.cancel(User.get_system_user(), silent=True)
+        booking.add_edit_log(ReservationEditLog(
+            user_name=User.get_system_user().full_name,
+            info=['Cancelled automatically due to lack of confirmation']
+        ))
+        notify_automatic_cancellation(booking)
+
+    db.session.flush()
+
+    # bookings which are about to be cancelled
+    for email in _build_query(2):
+        notify_about_to_cancel(email)
+    db.session.commit()
+
+
+@celery.task
+def update_access_permissions(booking, modify_dates=None):
```

Comment:
`old_dates` instead of `modify_dates` might be clearer (or is this where you mentioned on Zoom that you'd like to have a GET API on the ADaMS site to avoid having to pass it altogether?)

Reply:
I wanted to make it clear that this would entail an additional action, called `modify_dates`. `old_dates` might work too, although it's not an action.

Reply:
ah, good point... i guess two arguments would be the cleanest option (bool to indicate that you want to update, and the old dates in a separate arg), but at the same time it also means that passing just on argument would most likely always be wrong ;x

anyway, just naming, so could just leave it as it is right now as well ;)